[RNMobile] Add withIsConnected higher order component

derekblank opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
Adds a mobile connectivity status HOC.

* https://github.com/wordpress-mobile/gutenberg-mobile/pull/6457

## Why?
To be able to use the value returned from `useIsConnected` hook as a higher-order component.

## How?
Creates a higher-order component called `withIsConnected` that returns the value of `useIsConnected` to the WrappedComponent. 

## Testing Instructions
1. Apply the following patch:
    <details><summary>Test Diff</summary>

    ```diff
    diff --git a/packages/block-editor/src/components/media-upload-progress/index.native.js b/packages/block-editor/src/components/media-upload-progress/index.native.js
	index b64b08eec0..cf4b26f267 100644
	--- a/packages/block-editor/src/components/media-upload-progress/index.native.js
	+++ b/packages/block-editor/src/components/media-upload-progress/index.native.js
	@@ -10,6 +10,7 @@ import { Component } from '@wordpress/element';
	import { Spinner } from '@wordpress/components';
	import { __ } from '@wordpress/i18n';
	import { subscribeMediaUpload } from '@wordpress/react-native-bridge';
	+import { withIsConnected } from '@wordpress/compose';
	
	/**
	 * Internal dependencies
	@@ -32,6 +33,7 @@ export class MediaUploadProgress extends Component {
			};
	
			this.mediaUpload = this.mediaUpload.bind( this );
	+		this.getRetryMessage = this.getRetryMessage.bind( this );
		}
	
		componentDidMount() {
	@@ -115,19 +117,26 @@ export class MediaUploadProgress extends Component {
			}
		}
	
	+	getRetryMessage() {
	+		if ( this.props.isConnected === true ) {
	+			// eslint-disable-next-line @wordpress/i18n-no-collapsible-whitespace
	+			return __( 'Failed to insert media.\nTap for more info.' );
	+		}
	+		return __( 'You are currently offline.' );
	+	}
	+
		render() {
			const { renderContent = () => null } = this.props;
			const { isUploadInProgress, isUploadFailed } = this.state;
			const showSpinner = this.state.isUploadInProgress;
			const progress = this.state.progress * 100;
	-		// eslint-disable-next-line @wordpress/i18n-no-collapsible-whitespace
	-		const retryMessage = __(
	-			'Failed to insert media.\nTap for more info.'
	-		);
	+		const retryMessage = this.getRetryMessage();
	
			const progressBarStyle = [
				styles.progressBar,
	-			showSpinner || styles.progressBarHidden,
	+			showSpinner && this.props.isConnected
	+				? styles.progressBarHidden
	+				: null,
				this.props.progressBarStyle,
			];
	
	@@ -158,4 +167,4 @@ export class MediaUploadProgress extends Component {
		}
	}
	
	-export default MediaUploadProgress;
	+export default withIsConnected( MediaUploadProgress );
</details>

2. Run the host app with Gutenberg Mobile running locally
3. Enable airplane mode
4. Upload an image
5. Note that "You are currently offline." is displayed (instead of "Failed to insert media. Tap for more info.")

derekblank committed:
Add withIsConnected HOC

dcalhoun commented:
> 1. Apply the following patch:
>     Test Diff

For future reference, I encountered a `error: corrupt patch at line 6` error attempting to apply the diff. After manually recreating the diff, it appears that the diff file was missing preceding spaces. So, maybe the method used for copying/pasting the diff stripped these values. To illustrate the point, here is a meta diff of the diff. 

<details><summary>Image of diff changes</summary>

<img width="142" alt="Diff missing preceding spaces on unchanged lines" src="https://github.com/WordPress/gutenberg/assets/438664/43bbc324-5fe4-4616-828c-0a397d191309">

</details>

<details><summary>Diff of the Diff</summary>

```diff
diff --git a/packages/block-editor/src/components/media-upload-progress/index.native.js b/packages/block-editor/src/components/media-upload-progress/index.native.js
index b64b08eec0..4f3671dc7c 100644
--- a/packages/block-editor/src/components/media-upload-progress/index.native.js
+++ b/packages/block-editor/src/components/media-upload-progress/index.native.js
@@ -10,6 +10,7 @@ import { Component } from '@wordpress/element';
 import { Spinner } from '@wordpress/components';
 import { __ } from '@wordpress/i18n';
 import { subscribeMediaUpload } from '@wordpress/react-native-bridge';
+import { withIsConnected } from '@wordpress/compose';
 
 /**
  * Internal dependencies
@@ -32,6 +33,7 @@ export class MediaUploadProgress extends Component {
 		};
 
 		this.mediaUpload = this.mediaUpload.bind( this );
+		this.getRetryMessage = this.getRetryMessage.bind( this );
 	}
 
 	componentDidMount() {
@@ -115,19 +117,27 @@ export class MediaUploadProgress extends Component {
 		}
 	}
 
+	getRetryMessage() {
+		if ( this.props.isConnected ) {
+			// eslint-disable-next-line @wordpress/i18n-no-collapsible-whitespace
+			return __( 'Failed to insert media.\nTap for more info.' );
+		}
+
+		return __( 'You are currently offline.' );
+	}
+
 	render() {
 		const { renderContent = () => null } = this.props;
 		const { isUploadInProgress, isUploadFailed } = this.state;
 		const showSpinner = this.state.isUploadInProgress;
 		const progress = this.state.progress * 100;
-		// eslint-disable-next-line @wordpress/i18n-no-collapsible-whitespace
-		const retryMessage = __(
-			'Failed to insert media.\nTap for more info.'
-		);
+		const retryMessage = this.getRetryMessage();
 
 		const progressBarStyle = [
 			styles.progressBar,
-			showSpinner || styles.progressBarHidden,
+			showSpinner && this.props.isConnected
+				? styles.progressBarHidden
+				: null,
 			this.props.progressBarStyle,
 		];
 
@@ -158,4 +168,4 @@ export class MediaUploadProgress extends Component {
 	}
 }
 
-export default MediaUploadProgress;
+export default withIsConnected( MediaUploadProgress );
diff --git a/test.diff b/test.diff
index c66821a7bb..b29ba3085c 100644
--- a/test.diff
+++ b/test.diff
@@ -1,58 +1,59 @@
 diff --git a/packages/block-editor/src/components/media-upload-progress/index.native.js b/packages/block-editor/src/components/media-upload-progress/index.native.js
-index b64b08eec0..cf4b26f267 100644
+index b64b08eec0..4f3671dc7c 100644
 --- a/packages/block-editor/src/components/media-upload-progress/index.native.js
 +++ b/packages/block-editor/src/components/media-upload-progress/index.native.js
 @@ -10,6 +10,7 @@ import { Component } from '@wordpress/element';
-import { Spinner } from '@wordpress/components';
-import { __ } from '@wordpress/i18n';
-import { subscribeMediaUpload } from '@wordpress/react-native-bridge';
+ import { Spinner } from '@wordpress/components';
+ import { __ } from '@wordpress/i18n';
+ import { subscribeMediaUpload } from '@wordpress/react-native-bridge';
 +import { withIsConnected } from '@wordpress/compose';
-
-/**
- * Internal dependencies
+ 
+ /**
+  * Internal dependencies
 @@ -32,6 +33,7 @@ export class MediaUploadProgress extends Component {
-		};
-
-		this.mediaUpload = this.mediaUpload.bind( this );
+ 		};
+ 
+ 		this.mediaUpload = this.mediaUpload.bind( this );
 +		this.getRetryMessage = this.getRetryMessage.bind( this );
-	}
-
-	componentDidMount() {
-@@ -115,19 +117,26 @@ export class MediaUploadProgress extends Component {
-		}
-	}
-
+ 	}
+ 
+ 	componentDidMount() {
+@@ -115,19 +117,27 @@ export class MediaUploadProgress extends Component {
+ 		}
+ 	}
+ 
 +	getRetryMessage() {
-+		if ( this.props.isConnected === true ) {
++		if ( this.props.isConnected ) {
 +			// eslint-disable-next-line @wordpress/i18n-no-collapsible-whitespace
 +			return __( 'Failed to insert media.\nTap for more info.' );
 +		}
++
 +		return __( 'You are currently offline.' );
 +	}
 +
-	render() {
-		const { renderContent = () => null } = this.props;
-		const { isUploadInProgress, isUploadFailed } = this.state;
-		const showSpinner = this.state.isUploadInProgress;
-		const progress = this.state.progress * 100;
+ 	render() {
+ 		const { renderContent = () => null } = this.props;
+ 		const { isUploadInProgress, isUploadFailed } = this.state;
+ 		const showSpinner = this.state.isUploadInProgress;
+ 		const progress = this.state.progress * 100;
 -		// eslint-disable-next-line @wordpress/i18n-no-collapsible-whitespace
 -		const retryMessage = __(
 -			'Failed to insert media.\nTap for more info.'
 -		);
 +		const retryMessage = this.getRetryMessage();
-
-		const progressBarStyle = [
-			styles.progressBar,
+ 
+ 		const progressBarStyle = [
+ 			styles.progressBar,
 -			showSpinner || styles.progressBarHidden,
 +			showSpinner && this.props.isConnected
 +				? styles.progressBarHidden
 +				: null,
-			this.props.progressBarStyle,
-		];
-
-@@ -158,4 +167,4 @@ export class MediaUploadProgress extends Component {
-	}
-}
-
+ 			this.props.progressBarStyle,
+ 		];
+ 
+@@ -158,4 +168,4 @@ export class MediaUploadProgress extends Component {
+ 	}
+ }
+ 
 -export default MediaUploadProgress;
-+export default withIsConnected( MediaUploadProgress );
\ No newline at end of file
++export default withIsConnected( MediaUploadProgress );

```

</details>

dcalhoun commented on packages/compose/src/higher-order/with-is-connected/README.md:
```diff
@@ -0,0 +1,19 @@
+# withIsConnected
+
+`withIsConnected` provides a true/false mobile connectivity status based on the [useIsConnected hook](https://github.com/WordPress/gutenberg/blob/aadf8c66ff345b176041ad82e3793191ade5d271/packages/react-native-bridge/index.js#L193-L212).
```
Generally speaking, I think it is best to use permalinks, but in the case of documentation residing in this repository I think this would lead to outdated links. 

```suggestion
`withIsConnected` provides a true/false mobile connectivity status based on the `useIsConnected` hook found in the [bridge](https://github.com/WordPress/gutenberg/blob/trunk/packages/react-native-bridge/index.js).
```

dcalhoun commented on packages/compose/src/higher-order/with-is-connected/index.native.js:
```diff
@@ -0,0 +1,18 @@
+
+const withIsConnected = createHigherOrderComponent( ( WrappedComponent ) => {
+	return ( props ) => {
+		const isConnected = useIsConnected();
```
Noting that this needed to be destructured. I'll fix it in a commit. 
```suggestion
		const { isConnected } = useIsConnected();
```

dcalhoun committed:
fix: Destructure isConnected value from useIsConnected Hook

The previous value was always `undefined`.

dcalhoun committed:
docs: Avoid permalink in documentation

Permalinks would likely referenced outdated code at some point. Documentation is generally expected to be up-to-date.

dcalhoun commented:
> Testing Instructions

I was unable to get the provided testing diff work as expected in the app, even after manually applying the changes. The following diff represents the testing approach I took. 

<details><summary>Alternative Testing Diff</summary>

```diff
diff --git a/packages/block-library/src/paragraph/edit.native.js b/packages/block-library/src/paragraph/edit.native.js
index 98662ba21c..1ef8e4ff07 100644
--- a/packages/block-library/src/paragraph/edit.native.js
+++ b/packages/block-library/src/paragraph/edit.native.js
@@ -1,3 +1,8 @@
+/**
+ * External dependencies
+ */
+import { Text } from 'react-native';
+
 /**
  * WordPress dependencies
  */
@@ -11,6 +16,7 @@ import {
 } from '@wordpress/block-editor';
 import { useCallback } from '@wordpress/element';
 import { useSelect } from '@wordpress/data';
+import { withIsConnected } from '@wordpress/compose';
 
 const name = 'core/paragraph';
 
@@ -24,6 +30,7 @@ function ParagraphBlock( {
 	style,
 	clientId,
 	parentBlockAlignment,
+	isConnected,
 } ) {
 	const isRTL = useSelect( ( select ) => {
 		return !! select( blockEditorStore ).getSettings().isRTL;
@@ -52,6 +59,7 @@ function ParagraphBlock( {
 
 	const textAlignment = align || parentTextAlignment;
 
+	console.log( '>>>', { isConnected } );
 	return (
 		<>
 			<BlockControls group="block">
@@ -61,6 +69,11 @@ function ParagraphBlock( {
 					onChange={ onAlignmentChange }
 				/>
 			</BlockControls>
+			<Text>
+				{ typeof isConnected === 'boolean'
+					? isConnected.toString()
+					: 'null' }
+			</Text>
 			<RichText
 				identifier="content"
 				tagName="p"
@@ -101,4 +114,4 @@ function ParagraphBlock( {
 	);
 }
 
-export default ParagraphBlock;
+export default withIsConnected( ParagraphBlock );
```

</details>

dcalhoun committed:
fix: Avoid cyclic dependency

The test environment threw errors due to `@wordpress/compose` importing
itself.

dcalhoun committed:
refactor: Move react-native-bridge import to WordPress group

Fix lint error.
